### PR TITLE
fix: should not destroy streams

### PR DESCRIPTION
**Pull request by dead-horse**
use black-hole-stream to make sure stream's data has been read.

since user may set `this.body` to a http `IncomingMessage`, if we destroy this stream, it will destroy the http socket that make `keep-alive` failed, and may affect other http requests reusing this socket.
I think maybe we shouldn't destroy streams, only need to insure streams' data will be read and don't leak.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -47,14 +47,15 @@
...
     "should-http": "0.0.3",
     "supertest": "^1.0.1",
-    "test-console": "^0.7.1"
+    "test-console": "^0.7.1",
+    "urllib": "^2.5.0"
   },
   "engines": {
-    "node": ">= 0.12.0",
-    "iojs": ">= 1.0.0"
+    "node": ">= 0.12.0"
```
Comment (JacksonTian) [inline package.json]: node >= 4

Comment (dead-horse) [inline package.json]: can we drop the support of node < 4 in koa@1 ? /cc @tj @jonathanong

Comment (JacksonTian) [inline package.json]: oh, master is v1.

**Comment by dead-horse** [comment]
ping @jonathanong @fengmk2

**Comment by fengmk2** [comment]
LGTM

File: .travis.yml [hunk 2]
```diff
@@ -1,10 +1,8 @@
 language: node_js
 node_js:
   - "0.12"
-  - "1"
-  - "2"
-  - "3"
   - "4"
+  - "5"
```
Comment (juliangruber) [inline .travis.yml]: this should be its own pull request

File: package.json (hunk 1 above)
Comment (juliangruber) [inline package.json]: this should be its own pull request

File: test/application.js [hunk 3]
```diff
@@ -872,6 +876,69 @@ describe('app.respond', function(){
       .get('/')
       .expect(404, done);
     })
+
+    it('should insure stream do not leak', function(done){
```
Comment (juliangruber) [inline test/application.js]: insure => ensure

**Comment by dead-horse** [comment]
@juliangruber all fixed

File: test/application.js [hunk 4]
```diff
@@ -872,6 +876,69 @@ describe('app.respond', function(){
       .get('/')
       .expect(404, done);
     })
+
+    it('should ensure stream do not leak', function(done){
+      done = pedding(3, done);
+      var app = koa();
+      let stream1 = fs.createReadStream(__filename);
+      let stream2 = fs.createReadStream(__filename);
+      stream1.once('end', done);
```
Comment (juliangruber) [inline test/application.js]: i don't understand why this one would end as well? it's never read from

**Comment by juliangruber** [comment]
hmmm, this seems more like an edge case to me. are there other cases that would benefit from this change?

File: test/application.js (hunk 4 above)
Comment (dead-horse) [inline test/application.js]: we'll destroy(consume) all stream's that set to `this.body` here https://github.com/koajs/koa/pull/612/files#diff-d86a59ede7d999db4b7bc43cb25a1c11R166

**Comment by dead-horse** [comment]
The most serious problem now is http keepalive sockets will be destroyed, and we must fix this issue. Not sure any other side effect  if we destroy the streams.

**Comment by juliangruber** [comment]
are you sure destroying a http stream will affect other requests on the same socket? sounds more like a node bug to me

**Comment by dead-horse** [comment]
If you destroy an `IncommingMessage`, it will destroy the `socket` behind: https://github.com/nodejs/node/blob/master/lib/_http_incoming.js#L92

That will affect other `IncommingMessage`'s that reusing the same socket and will cause `socket hang up` error. This is reported in https://github.com/aliyun/oss-nodejs-sdk/issues/40 (sorry this issue is in chinese).

**Comment by dead-horse** [comment]
any suggestions? I'll merge this PR if no response in the next day. :)

**Comment by juliangruber** [comment]
Hey man, no need to put a deadline on it. If you need it urgently use a fork or find a userland solution.

**Comment by dead-horse** [comment]
I think this is just a bug fix. so if anyone doubt this patch, I'll leave it here and fix in our own framework for now. :)

**Comment by juliangruber** [comment]
your patch fails if the body stream is endless, like for example:

```js
this.body = fs.createReadStream('/dev/urandom');
```

or more real-worldy

```js
this.body = serverSentEvents(createUpdateStream());
```

in the cases above it's required to destroy the stream.

**Comment by juliangruber** [comment]
and trying to read it all out will never finish

**Comment by dead-horse** [comment]
although i don't think pipe an endless stream to response meaningful. but maybe there is a solution has less impact. detect the stream in `destroy` and don't destroy `IncommingMessage`(a breaking change in `destroy`).

**Comment by dead-horse** [comment]
@juliangruber how about this?

**Comment by juliangruber** [comment]
too edge-casy. koa should not need to worry about this.

Just to be sure, this would work for you, right?

```js
httpStream.destroy = blackHole.bind(null, httpStream);
this.body = httpStream;
```

**Comment by dead-horse** [comment]
it is a common usage that use koa as a http proxy, and anybody who use `this.body = yield request('http://foo.bar/'); //keep-alive` will hit this "edge case", and it is really hard to figure out why. so I think we should fix it in framework level, because it is a bug in koa, and we can't let everyone to use `httpStream.destroy = blackHole.bind(null, httpStream);` hack.

**Comment by jonathanong** [comment]
maybe a better solution is to just wrap the http stream in another stream so you never set the http stream as the actual `.body=`.

```js
const PassThrough = require('stream').PassThrough

app.use(function * (next) {
  this.body = httpStream.on('error', this.onerror).pipe(PassThrough())
})
```

**Comment by juliangruber** [comment]
> would need docs on this though... super edge case

what about documenting that a body stream can be `.destroy`ed and showing how to guard against that

**Comment by tejasmanohar** [comment]
>what about documenting that a body stream can be `.destroy`ed and showing how to guard against that
 Show all checks

+1 @juliangruber

File: lib/response.js [hunk 5]
```diff
@@ -161,8 +164,15 @@ module.exports = {
     }
 
     // stream
-    if ('function' == typeof val.pipe) {
-      onFinish(this.res, destroy.bind(null, val));
+    if (val instanceof Stream) {
+      onFinish(this.res, function(){
+        // don't destroy http IncomingMessage, keep `keep-alive` conncetion alive.
+        if (val instanceof http.IncomingMessage) {
+          if (val.readable) val.pipe(new BlackHoleStream());
```
Comment (jonathanong) [inline lib/response.js]: why do you need black hole stream? just `val.resume()` to dump it

**Comment by jonathanong** [comment]
anyways prefer going the documentation route. will close this when i or someone else adds documentation

**Comment by dead-horse** [comment]
It's ok to add some warning in documentation.

**Comment by halt-hammerzeit** [comment]
Didn't understand a thing from the documentation on what's happening here.
It's just like "warning, never set a stream as ctx.body".

**Comment by jiajianrong** [comment]
@catamphetamine, they were warning not to set a http-request-alike-stream which has keep-alive enabled. Because the framework will close the stream by default, which will break keep-alive feature. 
Keep-alive means to reuse tcp socket connection for two or more http connections.

Generally, it's OK with `ctx.body = fs-read-stream`. 
Just be careful with `ctx.body = request('host:port/path')`.

**Comment by rhinoceros** [comment]
@dead-horse  
cnpmjs.org中目前用的koa的版本是"koa": "^1.2.0", 如果使用oss作为nfs， 在从oss上get的时候会有一堆的sockets hang up。

有什么好的办法能避免这类错误吗？

https://github.com/cnpm/cnpmjs.org/issues/1401

谢谢

image

**Comment by zhangwilling** [comment]
@dead-horse  马哥好，昨天夜里见到这个 issue 忘了跟你打招呼了，白天来踩一下。

----- 骄傲的用中文交流
